File: Homework1/Hw1/Parser.cs
namespace Hw1;

public static class Parser
{
    public static void ParseCalcArguments(string[] args, 
        out double val1, 
        out CalculatorOperation operation, 
        out double val2)
    {
        if (!IsArgLengthSupported(args)) 
            throw new ArgumentException(
                "Недостаточно аргументов, их должно быть не менее трех.",
                nameof(args));

        val1 = ParseOperand(args[0]);
        val2 = ParseOperand(args[2]);
        
        operation = ParseOperation(args[1]);
        if (operation == CalculatorOperation.Undefined) 
            throw new InvalidOperationException(nameof(operation));
    }

    private static bool IsArgLengthSupported(string[] args) => args.Length == 3;

    private static CalculatorOperation ParseOperation(string arg)
    {
        return arg switch
        {
            "+" => CalculatorOperation.Plus,
            "-" => CalculatorOperation.Minus,
            "*" => CalculatorOperation.Multiply,
            "/" => CalculatorOperation.Divide,
            _ => CalculatorOperation.Undefined
        };
    }

    private static double ParseOperand(string argValue)
    {
        if (!double.TryParse(argValue, out var result)) 
            throw new ArgumentException("операнд не является числом.");

        return result;
    }
}

File: Homework4/Hw4/Parser.fs
module Hw4.Parser

open System
open Hw4.Calculator


type CalcOptions = {
    arg1: float
    arg2: float
    operation: CalculatorOperation
}

let isArgLengthSupported (args : string[]) = Array.length args = 3

let parseOperation (arg : string) =
    match arg with
    | "+" -> CalculatorOperation.Plus
    | "-" -> CalculatorOperation.Minus
    | "*" -> CalculatorOperation.Multiply
    | "/" -> CalculatorOperation.Divide
    | _ -> CalculatorOperation.Undefined

let parseDouble (arg: string) =
    let mutable result = 0.0
    match Double.TryParse(arg, &result) with
    | true -> result
    | false -> ArgumentException() |> raise
    
let parseCalcArguments(args : string[]) =
    match isArgLengthSupported args with
    | true -> match parseOperation args[1] with
              | CalculatorOperation.Undefined -> ArgumentException() |> raise
              | operation -> {
                  arg1 = parseDouble args[0]
                  arg2 = parseDouble args[2]
                  operation = operation
                  }
    | false -> ArgumentException() |> raise